Remove VariableCombos class and correctly handle two constant slot constraints

topnessman opened the pull request:
Encode two-constant constraints:
If the constraint holds returns emptyValue
Else return a contradictoryValue to ensure no solution exists

Remove VariableCombos class. Use more object-oriented way to implement this.

Early verification in ConstraintManager. If a two-constant constraint doesn't hold, exit immediately and abort the inference process.

Rollback to checker report and remove -Awarn from InferenceMain

CharlesZ-Chen commented on src/checkers/inference/solver/frontend/VariableCombos.java:
```diff
@@ -25,9 +26,13 @@
-    public VariableCombos(S emptyValue) {
+    public VariableCombos(S emptyValue, S contradictoryValue, Lattice lattice) {
         this.emptyValue = emptyValue;
+        this.contradictoryValue = contradictoryValue;
```
Mier, here are my opinion on this `contradictoryValue`, I vote an opposite on this change. Reasons:

1. Efficiency: we can actually early abort/exit if we found a constraint with constant slots is contradict. There is no need to passing a contradictory value into solver to enforce solver unsatisfiable.

2. Reasoning: if we substituted the contradictory constant constraint with an arbitrary contradictory value, then implicitly solver lost the reasoning of this unsatisfiable case ---- the minimized core set reported by solver would be an arbitrary contradictory value. --- even worse, what if this pre-set contradictory value is semantically duplicated with a real contradictory case? (i.e. what if some other cases generated the same core set with this arbitrary contradictory value). Then we cannot distinct if the reasoning represents a contradictory constant constraint, or other cases.

So here I suggest to remove this `contradictoryValue`, and just simply early exit with `Unsatisfiable` if we found a contradictory constant constraint.

CharlesZ-Chen commented on src/checkers/inference/solver/backend/SolverType.java:
```diff
@@ -38,14 +39,14 @@ private SolverType(String simpleName, Class<? extends SolverAdapter<?>> solverAd
         } catch (Exception e) {
             ErrorReporter.errorAbort(
-                    "Exception happens when creating default format translator for " + simpleName + " backend.", e);
+                    "Exception happens when creating default format t, ranslator for " + simpleName + " backend.", e);
```
"t, ranslator" -> "translator"

CharlesZ-Chen commented on src/checkers/inference/solver/backend/encoder/AbstractConstraintEncoder.java:
```diff
@@ -0,0 +1,32 @@
+    protected final Lattice lattice;
+    protected final ConstraintVerifier verifier;
+    protected final ConstraintEncodingT emptyValue;
+    // Clauses that are always false. Used to model unsatisfactory constraint between two
```
This code comment seems related to MaxSat encoding and is not general for AbstractConstraintEncoder.

topnessman replied:
Done.

topnessman commented:
This PR is a fix for issues: #78 and #84

topnessman commented:
@CharlesZ-Chen This PR is ready to do a final round of review. What needs to be done is adding domain specific contradictory value to LogiQL and Z3 backend.

topnessman commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -0,0 +1,161 @@
+
+    /**Remember to call this method at the end of subclass FormatTranslator constructor(MaxSAT and LogiQL) or at the last
+     * step of initializing FormatTranslator - in Z3BitVector backend, it's initSolver() method. Because the creation of
+     * encoders might not only need the parameters - lattice and verifier, but also need fields depending on concrete backend
```
Remove lattice from comment

CharlesZ-Chen commented on src/checkers/inference/InferenceMain.java:
```diff
@@ -154,8 +154,7 @@ private void startCheckerFramework() {
                 "-Xmaxerrs", "1000",
                 "-XDignore.symbol.file",
-                "-AprintErrorStack",
-                "-Awarns"));
```
What Is the reason of removing "-Awarns"?

topnessman replied:
Definition of "-Awarns" is here:
    // Treat checker errors as warnings
    // org.checkerframework.framework.source.SourceChecker.report
    "warns",

>     // Treat checker errors as warnings
    // org.checkerframework.framework.source.SourceChecker.report
    "warns",
Inference, it suppresses error of unsatisfiable constraint over two ConstantSlots into warning, and let compiler pass. As a result, we don't see the error message and inference doesn't exit directly.
After I removed this, I see the error message and inference exits directly when encountered unsatisfiable constraints.

CharlesZ-Chen replied:
Ok, got it. Thanks for the explanation!

CharlesZ-Chen commented on src/checkers/inference/solver/backend/encoder/binary/ComparableConstraintEncoder.java:
```diff
@@ -0,0 +1,7 @@
+/**
+ * Created by mier on 07/11/17.
+ */
+public interface ComparableConstraintEncoder<ConstraintEncodingT> extends BinaryConstraintEncoder<ConstraintEncodingT> {
```
One question: What is the benefit of having those empty interfaces for each different constraint type? There seems no difference to let concrete ConstraintEncoder directly implements `BinaryConstraintEncoder` or `CombineConstraintEncoder`, because this interface is empty and add no new APIs.

And also, for a new comer, when he/she looks a concrete ConstraintEncoder implementation, and trace up to one of those empty interfaces, it seems a bit confusing to him/her....

Can we simply remove those empty interfaces?

topnessman replied:
Interfaces that have empty methods are called marker interfaces. Its purpose is to tell compiler statically about the information about the instances of that interface.
Truly, Subtype, Equality, Inequality and ComparableConstraintEncoder interfaces doesn't define any new methods, but it successfully restricted what kind of instances when you override the factory methods to create ConstraintEncoders.
For example, we have two methods in AbstractFormatTranslator:
`    protected SubtypeConstraintEncoder<ConstraintEncodingT> createSubtypeConstraintEncoder(ConstraintVerifier verifier) {
        return null;
    }`
`    protected EqualityConstraintEncoder<ConstraintEncodingT> createEqualityConstraintEncoder(ConstraintVerifier verifier) {
        return null;
    }`
If you don't have that marker interfaces, you'll use BinaryConstraintEncoder as the return type for both methods. As a result, you can either pass "EqualityConstraintEncoder" in createSubtypeConstaintEncoder and mess up the four encoders in each other by accident. But if you have finer granularity using the marker interface, you are only allowed to pass the subtypeconstrainencoder in createSubtypeConstraintEncoder, but not in createEqualityConstraintEncoder, or the other two.

CharlesZ-Chen replied:
Marker interfaces usually used as a mark, so that some place will have a check on the mark and have some special process with that mark, e.g. a class have to implement  empty interface `Cloneable` to mark this class is cloneable. (Some place will have an `obj instance of Cloneable` check, and throws Exception if that `obj` does not implement `Cloneable`.

However, here these marker interfaces are used as Type Restrictions --- i.e. there were no special place to check those marks except subtype checking. Then it is actually no very helpful, as we gain no better safety for those marker interfaces. Let's illustrate in your use case: here you try to avoid sub-class passed wrong encoders for each factory method. Your argument is by having those marker interfaces, we can avoid passing a `EqualityEncoder` to a `createSubtypeEncoder` method. This is true. However, why a programmer could make this mistake?( Because he can see the method name is `createSubtypeEncoder`, and why does he will return a `EqualityEncoder`?)

Ok, let's just assume this kind of mistake might happen. Then, a similar mistake would be having a `xxxSubtypeEncoder implements SubtypeEncoder` by its actual implementation is about encoding `EqualityConstriant` (This is statically allowed, as the parent interface methods takes in a `BinaryConstraint`).

Then, your marker interface cannot prevent this similar mistake.

Why these two mistake are similar? As programmer provide a wrong implementation, even with given a clearly text information about what he should implement (i.e. return `EqualityEncoder` in `createSubtypeEncoder`, or implement `EqualityConstraint` encoding in a `SubtypeEncoder`).

So my conclusion is, let's trust programmer will provide correct implementation with a given clearly defined text information. As these marker interfaces actually doesn't prevent this mistake at the root.

```

topnessman replied:
> " have some special process with that mark".

`instance of Cloneable` check at runtime doesn't have black-box logic to test if an object is really clonable or not. It only "shallowly" checks an object is implementing Clonable interface or not. If a call to clone() method sees an object implementing the interface, it will pass the call. That special processing is now static. If a class doesn't implement correct encoder, then it is rejected.
 
> However, why a programmer could make this mistake?

By accident. I have experience that I mistyped something because of IDE auto completion. Sometimes it happens.

Marker interface is also a way to group together classes with similar functionality. Instead of putting all MaxSATSubtype, MaxSATEquality,  MaxSATInequality,  MaxSATComparable and the other eight all under BinaryConstraintEncoder, I like the feeling that they are grouped together by an interface that differentiate them. May I'm curious which solver backend doesn't support traslating preference constraint. If we have that marker interface, we can simply check who doesn't implement the PreferenceConstraintEncoder. It's crazily easier to look for it from at most three entries(each backend has one) than looking through at most 12 entries. If by accident, programmers are implementing wrong interface, e.g. EqualityEncoder implementing SubtypeEncoder interface, then it's also very easy to find it out when I investigate who implements SubtypeEncoder interface. But using only BinaryEncoder interface, I miss the second opportunity to catch this bug.
So overall, I like we have more control over what we're doing conceptually. Not just programmingly.

topnessman replied:
This is a screen shot : each class is listed clearly under its category.
![screenshot from 2017-11-09 10-35-12](https://user-images.githubusercontent.com/15336177/32613779-c474ceba-c539-11e7-954f-dd3418e4ccde.png)

CharlesZ-Chen replied:
Um... Grouping makes sense. Sure, having those markers interfaces do help us to group implementations. Thanks for the discussion!

topnessman replied:
Also thank you!

CharlesZ-Chen commented on src/checkers/inference/solver/backend/z3/Z3BitVectorFormatTranslator.java:
```diff
@@ -7,56 +7,42 @@
 
-    protected boolean isSubtypeSubSet() {
+    // TODO Charles, can this be public? Must it stay in Z3BitVectorFormatTranslator?
+    public boolean isSubtypeSubSet() {
```
I think this `isSubtypeSubSet` is only used in encoding subtype constraint. So maybe we can move it into Z3SubtypeConstraintEncoder. Let me try to do this.

topnessman replied:
Sure, feel free to refactor any Z3 specific staff if you think it's proper.

CharlesZ-Chen commented on src/checkers/inference/solver/backend/z3/encoder/Z3BitVectorSubtypeConstraintEncoder.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public BoolExpr encodeConstant_Variable(ConstantSlot subtype, VariableSlot supertype) {
+        BitVecExpr subtypeBv = z3BitVectorFormatTranslator.serializeConstantSlot(subtype);
```
I think maybe use `slot.serialize(z3BitVectorFormatTranslator)` would be better. Just add a comment as a reminder for myself. Let me refactor this.

topnessman replied:
Sure.

CharlesZ-Chen commented:
Thanks for this nice refactor! It solves the ugly inner class issues and now we have early verification!

Overall it looks good to me, and I only have two tiny comments. For Z3 specific, there were two refactors related to this change, so let me do this and merge into this PR (as them are strongly related to this PR, so it might be better to be a part of this PR also):

 - [ ] Move `isSubtypeSubSet()` from `Z3BitVectorFormatTranslator` to `Z3BVSubtypeConstraintEncoder`

 - [ ] Refactor the way of serializing slots: use `slot.serialize(z3BitVectorFormatTranslator)` instead

CharlesZ-Chen commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    @Override
+    public SlotEncodingT serialize(VariableSlot slot) {
```
Mier, since `AbstractFormatTranslator` is an abstract class, maybe we can remove those empty `serialize(Slot)` methods? Or this is your intention to set the default behavior of serializing slots should return null?

topnessman replied:
This is kind of default behaviour. I'm not quite clear about serializing slots. If we remove the default implementations, every subclass of FormatTranslator needs to implement those serialze(Slot) methods. Is this a normal scenario? If yes, we can remove them.

CharlesZ-Chen replied:
I think each concrete FormatTranslator need to implement serialize(Slot) methods, as it implements `Serializer` interface, which requires both serialize(Constraint)` and `serialize(Slot)` methods should be implemented.

For MaxSat and LogiQL, I think they have special problems with implementing `serialize(Slot)` methods, so maybe at general level we require each concrete FormatTranslator have to implement all serialize methods, and if some sub-class doesn't support some methods, make it explicit return null in that sub-class.

CharlesZ-Chen replied:
If we agree on this point, let me do this, as I currently working on this branch.

CharlesZ-Chen replied:
NVM, I just realized we already have default `null` implementation on `serialize(Constriant)`, so maybe just keep the consistence in general level, return `null` on all `serialize` methods.

topnessman replied:
Do you mean in MaxSAT and LogiQLFormatTranslator, all serialize(Slot) are explictily implemented to return null; In Z3FormatTranslator, serialize(Slot) methods are explicitly implemented with meaningful logic?

CharlesZ-Chen replied:
Yes, that is my original thought. But I just realize since we already have `null` implementation in `serialize(Constraint)`, where the design for sub-class is "only implement methods that it supports". So current `serialize(Slot)` return null makes sense, as it keep consistent with the design of `serialize(Constraint)`, i.e. a sub-class "only implement serialize methods that it supports".

So then we just need make this design documented in the abstract class.

topnessman replied:
Actually, in this part, I want to move encode(PreferenceConstraint) and encode(ExistentialConstraint) also into ConstraintEncoderDispatcher class but maybe using a new Name, e.g. ConstraintEncoderDirector

topnessman replied:
Sure, I'll document this design decision in the documentation.

CharlesZ-Chen replied:
Yes, that also improve the consistence. Agree on this point.

topnessman replied:
This encapsulation is done.

CharlesZ-Chen commented:
I've refactored z3 encoders, and add the real contradictory value for z3 encoders.

Now this PR looks good to me, just wait for the documentations.

Thanks a lot for the great work, Mier!

topnessman commented:
Thank you Charles for your great help, too! I'll add documentation today.

topnessman commented:
@CharlesZ-Chen @wmdietl 
Hi Charles,
I just pushed a commit that contains documentation and cleanup. If you like, you can go over the documentation quickly. I double checked it already so hopefully there won't be too much problems.

CharlesZ-Chen commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -24,15 +24,18 @@
     protected final Lattice lattice;
-    // verifier is private because subclasses doesn't use it. It only needs to be passed
-    // as an argument to encoder factory methods
+    /* verifier is private because subclasses doesn't use it. It only needs to be passed
```
Instead of explaining the visibility modifiers, document what this field is (this would be more helpful for others read your code, as they want to know what this field is about).

E.g.
```java
/**
* A verifier that verify if constraints between constant slots satisfiable.
*
* This verifier helps in optimizing encoding of constraints between constant slots.
* E.g. Instead of directly encode a constant constraint, we encode an empty value if
* it passed verification by this verifier, otherwise encode a contradictory value for it.
*/
private final ConstraintVerifier verifier;
```

topnessman replied:
Personally, I prefer having no documentation on fields at all. For me, you can jump to class Lattice and see the documentation there. If we document fields lattice and verifier, it looks duplicate for me. Do you think having documentation just because of it is a field of another class would help? If so, I can change it to rigorous javadoc.

CharlesZ-Chen commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -24,15 +24,18 @@
         implements FormatTranslator<SlotEncodingT, ConstraintEncodingT, SlotSolutionT>{
 
-    // lattice is protected because every subclass of AbstractFormatTranslator uses it.
+    /* lattice is protected because every subclass of AbstractFormatTranslator uses it.*/
```
Same, Instead of explaining the visibility modifiers, document what this field is. For lattice, we can simply document it as something like:
```
/**
* Target lattice for the type system.
*/

CharlesZ-Chen commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -24,15 +24,18 @@
 /**
- * Created by mier on 08/11/17.
+ * Abstract implementation of FormatTranslator that all concrete FormatTranslator should extend from.
+ * It has the necessary lattice that will be used and already setup the abstract structure and provided
```
Better to clarify what "abstract structure" it setup.

Also, maybe make it explicit here that if a subclass doesn't override a serialize method here, by default it assumes this sub-class does not support corresponding constraint/slot.

topnessman replied:
"abstract structure" really means the template of what a FormatTranslator looks like. But it looks too cumbersome for me to say, setup the seven encoder fields, factory methods that instantiate them and provides defaultImplementation etc.
I think the second point is reflected in the following documentation, isn't it?

CharlesZ-Chen commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -48,9 +51,9 @@ public AbstractFormatTranslator(Lattice lattice, ConstraintVerifier verifier) {
-     * encoders might not only need the parameters - lattice and verifier, but also need fields depending on concrete backend
+     * encoders might not only need the parameter - verifier, but also need fields depending on concrete backend
      * So after those dependant fields are initialized, call this method to finish initializing encoders*/
     protected void postInit() {
```
javadoc usually starts with one or two sentences, explain the purpose/main functionalities of this method/field. Then might have a separate paragraph for detail explanations of programming guide, conventions, etc.

Also, if this method is specific for post initializing encoders, rename it to 'postInitEncoders()`. (And it seems this is just the place initializing encoders? Maybe just call it `initEncoders()` as we don't have another place initializing encoders prior than this place...)

Better to refactor it to something like:

```java

/**
* Initialize concrete encoders that used by this format translator.
*
* It is sub-class responsibility to decide how to construct encoders, and 
* when is a proper timing to initialize encoders.
*
* E.g. For MaxSat and LogiQL, it initialize encoders with additional informations other than verifiers,
* and initialize them at the last step of it's FormatTranslator constructors.
*/
 protected void postInit() {
...
}
```

CharlesZ-Chen replied:
Also one thought: it seems this `postInit` method doesn't get called in this abstract class. This is seems a bad style: define a method implementation in an abstract class, but actually abstract class doesn't use it in any template it defines.

Could we enforce this method get called by the constructor in this abstract class? (or remove this method, just call all `createXxxEncoder()` methods in the constructor.)

For Z3, I can have a `setContext(context)` for each concrete encoder, and just set the context in `Z3FormatTranslator` constructor after the super constructor initialized those encoders.

CharlesZ-Chen commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -24,15 +24,18 @@
- * Created by mier on 08/11/17.
+ * Abstract implementation of FormatTranslator that all concrete FormatTranslator should extend from.
+ * It has the necessary lattice that will be used and already setup the abstract structure and provided
+ * default implementations so that you only need to override minimum methods your FormatTranslator is
```
Avoid using personal pronouns (you, we, us..) in code documents as at the most of times they are ambiguous.

Substitute personal pronouns with the clearly defined entities, e.g.

instead of:

"so that you only need to override minimum methods your FormatTranslator is interested."

refactor it to something like:

"so that sub-class of this abstract formatTranslator only need to override minimum methods that it is interested."

CharlesZ-Chen commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -91,9 +94,10 @@ protected void postInit() {
+    /** Caution: if subclass overrides the serialize(XXXConstraint) methods, nullness of the corresponding encoder
+     * field must be checked before dereferencing unless you overrode corresponding createXXXConstraintEncoder to
+     * return a non-null value. Because subclass might not support some constraint encoding and the encoders of those
+     * unsupported constraints are by default null. By default, serialize(XXXConstraint) methods don't encode constraints*/
```
This javadoc seems not specific to `serialize(SubtypeConstraint)`, maybe merge this one with the javadoc of this class?

CharlesZ-Chen commented on src/checkers/inference/solver/backend/AbstractFormatTranslator.java:
```diff
@@ -136,6 +140,8 @@ public ConstraintEncodingT serialize(ExistentialConstraint constraint) {
                 ConstraintEncoderCoordinator.redirect(constraint, existentialConstraintEncoder);
     }
 
+    /** Override serialize(XXXSlot) if your FormatTranslator subclass has a concrete logic to serialize
```
Same comment. This javadoc seems not specific to `serialize(VariableSlot)`.

CharlesZ-Chen commented on src/checkers/inference/solver/backend/encoder/ConstraintEncoderCoordinator.java:
```diff
@@ -12,7 +12,7 @@
 
 /**
- * Created by mier on 08/11/17.
+ * A coordinator class that has the logic how each encoder encodes its supported constraint
```
This seems dispatch constraints to the correct encoders based on the slot composition of the given constraint. So it is not "has the logic of how each encoder encodes constraint", but is "dispatch to the correct encoder to encode constraints".

CharlesZ-Chen commented on src/checkers/inference/solver/backend/encoder/SlotSlotCombo.java:
```diff
@@ -3,9 +3,14 @@
 
 /**
- * Created by mier on 06/11/17.
+ * Enum that models combination of slots so that encoders knows which encodeXXX() method should be
```
"so that encoders knows which encodeXXX() method" -> "so that EncoderCoordinator knows which encodeXXX() method of encoders it should dispatch to".

CharlesZ-Chen commented on src/checkers/inference/solver/backend/maxsat/encoder/MaxSATSubtypeConstraintEncoder.java:
```diff
@@ -8,25 +8,73 @@
+     * for subtype constraint, if supertype is constant slot, then the subtype
+     * cannot be the super type of supertype, same for subtype
+     *
+     * @param mustNotBe
```
If we don't have an explanation of a `@param`, remove it. (Or a better way would be add a simple explanation of this `@param`. The same as `@return`.

CharlesZ-Chen commented on src/checkers/inference/solver/backend/maxsat/encoder/MaxSATSubtypeConstraintEncoder.java:
```diff
@@ -8,25 +8,73 @@
+     * @param unknownType
+     * @param maybeSet
+     * @return
+     */
```
Remove this empty javadoc. Or fill it with some information.

CharlesZ-Chen commented on src/checkers/inference/util/ConstraintVerifier.java:
```diff
@@ -1,15 +1,16 @@
- * ConstraintManager and also passes the reference to this class to solver engine
- * so that FormatTranslator can use this class to verify two-constants constraint
- * in order to encode it more efficiently.
+ * This class verifies constraint between two ConstantSlots holds or not. We use this class
```
Avoid use personal pronoun "we". Clarify who are the potential client of this class.

CharlesZ-Chen commented:
@topnessman  Mier, thanks for adding the documentations! I've had a initial round of reviewing on your javadocs. Please see the comments. Any discussions are welcome!

CharlesZ-Chen commented:
Thanks for the improvement! Looks good to me, except the Z3 adaptation (I believe the `context` would be `null` in `Z3xxxEncoderFactory`.) Let me refactor Z3 and merge to this branch.

topnessman commented:
Yes, so I called createXXXEncoder()s again in initSolver() method.
I accidently mixed some errors into the PR which caused the build failure. Let me fix it first and you can add commit later to avoid unnecssary conflict.

CharlesZ-Chen commented:
> Yes, so I called createXXXEncoder()s again in initSolver() method.

Ok, let me think about how to refactor this.

> I accidently mixed some errors into the PR which caused the build failure. Let me fix it first and you can add commit later to avoid unnecssary conflict.

Sure.

topnessman commented:
I added back postinitialization logic(now called finishInitializingEncoders()) because you must wait for subclass constructors to initialiaze some dependant fields. 
I also updated the Javadoc a lot that I hope will be much more useful than the old one.
@CharlesZ-Chen @wmdietl

topnessman commented:
@CharlesZ-Chen I'm going to merge this PR if you don't have additional comments. Do you?

CharlesZ-Chen commented:
No, this PR looks good to me. I can have a separate PR for Z3 encoders
afterward.

Mier Ta <notifications@github.com>于2017年11月14日 周二18:48写道：

> @CharlesZ-Chen <https://github.com/charlesz-chen> I'm going to merge this
> PR if you don't have additional comments. Do you?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/opprop/checker-framework-inference/pull/96#issuecomment-344438902>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHWCda7DgqRMdzizsVF7Xc-fzJYT-qy1ks5s2ibmgaJpZM4QTgl0>
> .
>

topnessman commented:
Cool, thanks.
